crates/storage/src/connection/state_update.rs

```diff
@@ -464,21 +464,25 @@ impl Transaction<'_> {
     }
 
     /// Items are sorted in descending order.
-    pub fn state_diff_lengths(
+    pub fn state_diff_lengths_and_commitments(
         &self,
         start: BlockNumber,
         max_num_blocks: NonZeroUsize,
-    ) -> anyhow::Result<Vec<usize>> {
+    ) -> anyhow::Result<Vec<(usize, StateDiffCommitment)>> {
         let mut stmt = self
             .inner()
             .prepare_cached(
-                r"SELECT state_diff_length FROM block_headers WHERE number >= ? ORDER BY number ASC LIMIT ?",
+                r"SELECT (state_diff_length, state_diff_commitment) FROM block_headers WHERE number >= ? ORDER BY number ASC LIMIT ?",
```

Comment:
Have you actually tried running that query? It's invalid SQL syntax:

```sqlite> SELECT (state_diff_length, state_diff_commitment) FROM block_headers WHERE number=1;
Parse error: row value misused
```

```suggestion
                r"SELECT state_diff_length, state_diff_commitment FROM block_headers WHERE number >= ? ORDER BY number ASC LIMIT ?",
```